examples/hsm-example.cpp

```diff
@@ -26,20 +28,202 @@
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
+
+std::vector<uint8_t> ecdsaSign(const AsymmetricPrivateKey &privKey,
+                               const DigestTypes digestType,
+                               const ECDSASignatureFormat sigFormat,
+                               const std::vector<uint8_t> &message)
+{
+    std::shared_ptr<MessageSignatureCtx> signCtx;
+    try {
+        /* The argument hashFunction is optional. Default is SHA256
+         * The default signature format is ECDSASignatureFormat::ASN1_SEQUENCE_OF_INTS */
+        signCtx = std::make_shared<ECDSASignaturePrivateKeyCtx>(privKey, digestType, sigFormat);
+    } catch (MoCOCrWException &e) {
+        std::cerr << "Please check your ECC key. Failure creating context." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+
+    std::vector<uint8_t> signature;
+    try {
+        signature = signCtx->signMessage(message);
+    } catch (const MoCOCrWException &e) {
+        /* Possible reasons:
+         * - error in openssl (sign, padding, ...)
+         * - Hash function's digest size doesn't match the message's digest size
+         * - Invalid signature format set
+         */
+        std::cerr << "Failure occurred during signing." << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+    return signature;
+}
+
+void ecdsaVerify(const AsymmetricPublicKey &pubKey,
+                 const DigestTypes digestType,
+                 const ECDSASignatureFormat sigFormat,
+                 const std::vector<uint8_t> &signature,
+                 const std::vector<uint8_t> &message)
+{
+    std::shared_ptr<MessageVerificationCtx> verifyCtx;
+    try {
+        verifyCtx = std::make_shared<ECDSASignaturePublicKeyCtx>(pubKey, digestType, sigFormat);
+    } catch (const MoCOCrWException &e) {
+        std::cerr << "Please check your ECC key. Failure creating context." << std::endl;
+        std::cerr << e.what();
+        exit(EXIT_FAILURE);
+    }
+    try {
+        verifyCtx->verifyMessage(signature, message);
+    } catch (const MoCOCrWException &e) {
+        /* Possible reasons:
+         * - error in openssl (sign, padding, ...)
+         * - Hash function's digest size doesn't match the message's digest size
+         * - Invalid signature format set
+         * - Signature can't be parsed
+         * - Invalid signature
+         */
+        std::cerr << "Verification failed!" << std::endl;
+        std::cerr << e.what() << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
+
 int main(void)
 {
     // Information for engine loading and key management.
     std::string id("pkcs11");
     std::string modulePath("/usr/lib/softhsm/libsofthsm2.so");
     std::string pin("1234");
-    std::string keyID("1001");
-
-    // Step 2: Initialise the PKCS11 Engine.
+    std::vector<uint8_t> message = utility::fromHex("deadbeef");
     HsmEngine hsmEngine(id, modulePath, pin);
```

Comment:
hm, the pin seems to be associated to the HSM, instead of the key itself

hypothetical scenario... how could an `application A` create a key in the HSM, and share it only with an `application B`, such that no other applications (even some that use the HSM as well) can view it?

it does not seem possible with this mococrw design, unless I'm missing something

Reply:
I discussed the same thing with Ivan today. We'll move the "token" also into the HSMEngine. Then one HSMEngine represents on PKCS#11 token instead of the whole HSM. Then everything that belongs together token/PIN is in the same thing. This is not really clear from the current design.

Reply:
This will be handled in a separate PR. The scenario you described currently doesn't really work as you would expect.